### ISSUE TITLE
sendWithPromise("GetPage", ... null causing an error

### ISSUE THREAD

**Issue by jdc-cunningham** [wojtekmaj/react-pdf#974]
### Before you start - checklist

- [X] I followed instructions in documentation written for my React-PDF version
- [X] I have checked if this bug is not already reported
- [X] I have checked if an issue is not listed in [Known issues](https://github.com/wojtekmaj/react-pdf/wiki/Known-issues)
- [X] If I have a problem with PDF rendering, I checked if my PDF renders properly in [PDF.js demo](https://mozilla.github.io/pdf.js/web/viewer.html)

### Description

I use two instances of react-pdf. The main viewer is using a functional component and the second one is a class-based that's not visible. It's a background renderer to produce pdf thumbnail previews.

After recently pulling down **5.7.2** my background PDF render throws an error just before it renders. After reverting to 5.7.1 it's working fine.

- load 1 PDF it's fine
- load another PDF it breaks

The background rendering script is class-based, it is not visible on the DOM and is used for rendering the first page/producing preview images for PDF thumbnails. This has been working for several months/since 2021.

I put error handlers on the `<Document...` and `<Page..`. I have `onSourceError`, `onLoadError`, `onRendeError` handlers but those don't fire, the error above happens.

I will continue investigating/work on reproducible code unfortunately what I'm working on is not public.

Attached below are the images of the error.

### Steps to reproduce

Setup a class-based react-pdf component.

Pass down a file path (by props), it renders.

Pass down another file path

### Expected behavior

The PDFs render as the sources are updated.

### Actual behavior

Fails to render the PDF/produces the errors below

![err-1](https://user-images.githubusercontent.com/11015467/161202029-c1b87732-cee4-4b46-94b8-a0565df31c01.PNG)

![err-2](https://user-images.githubusercontent.com/11015467/161202035-85bdad00-7bd0-4d9b-ad3d-0d57f3cd7932.PNG)

### Environment

- **Browser (if applicable)**:
- **React-PDF version**:
- **React version**:
- **Webpack version (if applicable)**:

**Comment by jdc-cunningham** [comment]
Thanks for the quick response.

I loaded it up on a staging site/back to 5.7.2. This shows a gap in our testing to test this far (more than one document loaded in the background).

I did look to see if there's a direct call relevant there. I added an expanded callstack trace attached below, I think your hunch is right.

It could be that I'm using it wrong too. Here's a [gist](https://gist.github.com/jdc-cunningham/14dff0bd9c79b43dbb5d87c1b01d13a1) on the background renderer class-based component I have. This is a child of another component managing its state so the active file comes down as a prop. Also the functional instance is a sibling/same depth as this functional one.

Maybe it's the async `getDocPageImg` call?

![expanded-stack](https://user-images.githubusercontent.com/11015467/161288542-90ec1599-2b44-447c-b961-f891fc0b251a.png)

**Comment by jdc-cunningham** [comment]
Not a solution but I've capped the `react-pdf` version in my project to `5.7.1` for the time being.

**Comment by PerryAJ** [comment]
We've hit the same situation with an application using react 16.14.  Was able to workaround by rolling back to 5.7.1 as others have.  Don't have much else to add to previous reports, but did want to note that the root cause was one of a number of different errors on different subcomponents from Page down.  For future searchability, figured it was worth adding one example from a dev bundle created with 5.7.2:

```
Failed prop type: The prop 'page.commonObjs._objs' is marked as required in 'PageCanvasInternal', but its value is 'undefined'. Failed prop type: The prop 'page.commonObjs._objs' is marked as required in 'TextLayerInternal', but its value is 'undefined'
```

**Comment by FabricioK** [comment]
I`m having the same problem. Is this being addressed?

**Comment by wojtekmaj** [comment]
Does anyone have a reproducible CodeSandbox repo? I'm especially interested if this is still reproducible in v6.

**Comment by ldiqual** [comment]
We're having the same issue on v6.2.0. Reverting to v5 fixes it. We haven't been able to reproduce locally but it does happen in production and we're getting many errors. We re-render with different files many times during the component's lifetime.

<img width="1242" alt="image" src="https://user-images.githubusercontent.com/941453/208211051-aca10fe1-a631-4f74-9e1a-9cf6ac494ead.png">

<img width="955" alt="image" src="https://user-images.githubusercontent.com/941453/208211104-41f73562-1d6e-4700-8705-df68385dc83e.png">

**Comment by illusdolphin** [comment]
I think we solved the issue. It's mostly in the user-code together with internal logic of component.
Demo code is simple and does not handle file change situation. 
Solving for us was easy - just make page number of pages to render to 0 at the moment file is changed. 
Sample code below (not sure about 100% compiling, but the idea should be clear.
For sure this might be solved on component level, then it needs to ignore pages when loading is in progress.

`import React, { useState } from 'react';
import { Document, Page } from 'react-pdf';

function MyApp() {
  const [numPages, setNumPages] = useState(0);
  const [fileName, setFileName] = useState('somefile.pdf');
  const [pageNumber, setPageNumber] = useState(1);

  function onDocumentLoadSuccess({ numPages }) {
    setNumPages(numPages); // <== here is the issue #1 - we save page number of file, but in case it's changed, we know about new number only after it's loaded, before doc is loaded, this value is wrong
  }

  function setNewFile(){
    //file change event
    setNumPages(0);
    setFileName('newfile.pdf');
  }

  return (
    <div onClick={setNewFile}>
      <Document file={fileName} onLoadSuccess={onDocumentLoadSuccess}>
        {Array.from(new Array(numPages), (_, index) => (
                            <Page
                                key={`page_${index + 1}`}
                                pageNumber={index + 1}
                            />
                        ),
        )} { /* here we render all pages of file, but if new file is being loaded, we try loading pages that are not available */ }
      </Document>
      <p>
        Page {pageNumber} of {numPages}
      </p>
    </div>
  );
}`

**Comment by ldiqual** [comment]
I found a related issue in react-pdf-viewer that was due to reusing a destroyed worker:
 - Issue: https://github.com/react-pdf-viewer/react-pdf-viewer/issues/417
 - Fix: https://github.com/react-pdf-viewer/react-pdf-viewer/pull/424
It seems that we're encountering this bug when changing loading blob A then blob B very quickly. My guess is that when page 1 of blob A gets to load, the worker for blob A is destroyed which causes this error.

**Comment by pedroferreira37** [comment]
Just add useAsync or your own implementation when generating the url, the @ldiqual response is correctly, it changes from a to b very quickly and the worker is not setted. 

Also, if you gonna using some reactivity, i recommend using debounce in function that change url also

**Comment by MattL75** [comment]
Weird. I've only now started getting this error when trying to migrate to v7.

**Comment by cattale93** [comment]
Hello, I encountered the exact same problem after updating from v6.2.2 to v7.1.1.  I solved it controlling the page rendering. I allow my app to render the page only if the document has been successfully loaded (onDocumentLoadSuccess).

**Comment by WoldemortRules** [comment]
hey - can you add an example how did you resolved it ?@cattale93

**Comment by luxiangqiang** [comment]
The error can be temporarily bypassed by using:

```
useEffect(() => {
    setTimeout(() => {
      setIsRenderPdf(true)
    }, 500)
  }, [])
```

```
isRenderPdf ? 
<Document 
  key={pdfUrl} 
  file={pdfUrl} >
</Document>
```

**Comment by cattale93** [comment]
> hey - can you add an example how did you resolved it ?@cattale93

Hello @WoldemortRules 

```javascript
<Document
      className={'PDFDoc'}
      file={fileLocation}
      onLoadSuccess={this.onDocumentLoadSuccess}  // the pdf is rendered only based on a specific state in this function.
      onSourceError={this.onLoadError}
      onLoadError={this.onLoadError}
      renderMode={'canvas'}
      onItemClick={this.goToPage}
      options={{ cMapUrl: 'cmaps/',  cMapPacked: true,}}
>
```
According to the documentation _onLoadSuccess_ -> Function called when the document is successfully loaded.

**Comment by MasNotsram** [comment]
Same issue here with v6.0.3. Any news on a fix for this?

We can do conditional rendering, but as this worked in V5, it would be nice if this worked in React 18 versions of react-pdf.

**Comment by jiangxiaoqiang** [comment]
I am facing the same issue when using 7.3.3.

**Comment by little-buddy** [comment]
It seems I solved the problem. 
You only need to prevent Document rerender. 
The above reason is because rerender causes pdfjs to reload pdf with an error.

**Comment by huypham1411** [comment]
I only got this error when I resized the screen using the responsive dev tool, here is my call stack
```
Call Stack
WorkerTransport.getOptionalContentConfig
node_modules\pdfjs-dist\build\pdf.js (2481:0)
PDFPageProxy.render
node_modules\pdfjs-dist\build\pdf.js (1385:0)
drawPageOnCanvas
node_modules\react-pdf\dist\esm\Page\PageCanvas.js (68:0)
commitHookEffectListMount
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (20002:0)
commitHookPassiveMountEffects
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22055:0)
commitPassiveMountOnFiber
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22160:0)
recursivelyTraversePassiveMountEffects
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
commitPassiveMountOnFiber
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22271:0)
recursivelyTraversePassiveMountEffects
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
commitPassiveMountOnFiber
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22271:0)
recursivelyTraversePassiveMountEffects
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
commitPassiveMountOnFiber
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22157:0)
recursivelyTraversePassiveMountEffects
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
commitPassiveMountOnFiber
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22271:0)
recursivelyTraversePassiveMountEffects
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
commitPassiveMountOnFiber
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22271:0)
recursivelyTraversePassiveMountEffects
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
commitPassiveMountOnFiber
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22157:0)
recursivelyTraversePassiveMountEffects
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
commitPassiveMountOnFiber
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22271:0)
recursivelyTraversePassiveMountEffects
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
commitPassiveMountOnFiber
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22271:0)
recursivelyTraversePassiveMountEffects
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
commitPassiveMountOnFiber
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22157:0)
recursivelyTraversePassiveMountEffects
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
commitPassiveMountOnFiber
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22271:0)
recursivelyTraversePassiveMountEffects
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
commitPassiveMountOnFiber
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22271:0)
recursivelyTraversePassiveMountEffects
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
commitPassiveMountOnFiber
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22157:0)
recursivelyTraversePassiveMountEffects
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
commitPassiveMountOnFiber
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22157:0)
recursivelyTraversePassiveMountEffects
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
commitPassiveMountOnFiber
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22271:0)
recursivelyTraversePassiveMountEffects
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
commitPassiveMountOnFiber
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22157:0)
recursivelyTraversePassiveMountEffects
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
commitPassiveMountOnFiber
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22271:0)
recursivelyTraversePassiveMountEffects
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
commitPassiveMountOnFiber
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22157:0)
recursivelyTraversePassiveMountEffects
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
commitPassiveMountOnFiber
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22157:0)
recursivelyTraversePassiveMountEffects
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
commitPassiveMountOnFiber
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22157:0)
recursivelyTraversePassiveMountEffects
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
commitPassiveMountOnFiber
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22157:0)
recursivelyTraversePassiveMountEffects
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
commitPassiveMountOnFiber
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22271:0)
recursivelyTraversePassiveMountEffects
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
commitPassiveMountOnFiber
node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22157:0)
```

**Comment by jhunexjun** [comment]
My issue was solved by @illusdolphin.
My exact error is not the as sendWithPromise() but 
`pdf Cannot read properties of null (reading 'sendWithStream')`
But he gave me an idea. So here's the solution:
```
function onDocumentLoadSuccess(nextNumPages) {
  setNumPages(nextNumPages ?? 0);
}
```

So you just test the nextNumPage.

**Comment by name-undef1ned** [comment]
# step 1
 `  <Document
       key={1}
      file={playUrl}
      onLoadSuccess={onDocumentLoadSuccess}`
# step 2
` const onDocumentLoadSuccess = (...args) => {
    const info = args[0];
    setnumPages(info?.numPages??0)
    setIsRenderPdf(true)//设置状态
    
  }`
# step 3
`   {isRenderPdf&&new Array(10).fill('').map((cur, index) => (
            <Page
              key={index}
              pageNumber={index + 1}
              renderAnnotationLayer={false}
              renderTextLayer={false}
              width={window.innerWidth}
              renderMode='canvas'
            />
          ))}
 `

**Comment by Puetz** [comment]
For me it worked after simply wrapping my Pdf Component with React.memo.

`export default React.memo(PdfViewerComponent)`

**Comment by TJohn2017** [comment]
For anybody still struggling, I got it to work on our side by using the method from [this comment](https://github.com/wojtekmaj/react-pdf/issues/1473#issuecomment-1838145563) on a closed duplicate of this issue.

I'm using Next.js

`cp node_modules/react-pdf/node_modules/pdfjs-dist/build/pdf.worker.js public/pdf.worker.js`

Then, within your app where you'd like to setup the worker:

`import { pdfjs } from 'react-pdf';

pdfjs.GlobalWorkerOptions.workerSrc = "/pdf.worker.js"`

Honestly, I'm not totally sure why this worked. So, if anybody has a better understanding definitely feel free to reply. I'd be curious. Hope this helps!

**Comment by huypham1411** [comment]
> For anybody still struggling, I got it to work on our side by using the method from [this comment](https://github.com/wojtekmaj/react-pdf/issues/1473#issuecomment-1838145563) on a closed duplicate of this issue.
> 
> I'm using Next.js
> 
> `cp node_modules/react-pdf/node_modules/pdfjs-dist/build/pdf.worker.js public/pdf.worker.js`
> 
> Then, within your app where you'd like to setup the worker:
> 
> ```
> import { pdfjs } from 'react-pdf';
> 
> pdfjs.GlobalWorkerOptions.workerSrc = "/pdf.worker.js"
> ```
> 
> Honestly, I'm not totally sure why this worked. So, if anybody has a better understanding definitely feel free to reply. I'd be curious. Hope this helps!

Hi, I'm using Nextjs too and I also read the relative post about this issue but I don't understand this part
> Copy react-pdf/node_modules/pdfjs-dist/build/pdf.worker.js to public

the react-pdf is inside the node_modules so this make me a bit confuse, hope you can help me, appreciate

**Comment by TJohn2017** [comment]
> > For anybody still struggling, I got it to work on our side by using the method from [this comment](https://github.com/wojtekmaj/react-pdf/issues/1473#issuecomment-1838145563) on a closed duplicate of this issue.
> > I'm using Next.js
> > `cp node_modules/react-pdf/node_modules/pdfjs-dist/build/pdf.worker.js public/pdf.worker.js`
> > Then, within your app where you'd like to setup the worker:
> > ```
> > import { pdfjs } from 'react-pdf';
> > 
> > pdfjs.GlobalWorkerOptions.workerSrc = "/pdf.worker.js"
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Honestly, I'm not totally sure why this worked. So, if anybody has a better understanding definitely feel free to reply. I'd be curious. Hope this helps!
> 
> Hi, I'm using Nextjs too and I also read the relative post about this issue but I don't understand this part
> 
> > Copy react-pdf/node_modules/pdfjs-dist/build/pdf.worker.js to public
> 
> the react-pdf is inside the node_modules so this make me a bit confuse, hope you can help me, appreciate

So, this method didn't actually end up resolving my issue. It made it less likely to occur, and we didn't see it for a bit but since making this post I've had the same issue a handful more times now. If anybody is able to find a true solution I'd really appreciate it.

If you'd still like to try that method you should see that react-pdf has a node_modules inside of it, despite it being inside of your project's node_modules. The full filepath should be `node_modules/react-pdf/node_modules/pdfjs-dist/build/pdf.worker.js` from the base of your Nextjs project.

**Comment by Adermi** [comment]
我在添加自定义cMapUrl的时候也报这个错，之后把option给缓存就不会有这个问题了。

``` javascript
  const opt = useMemo(() => {
    return {
      cMapUrl: '/bcmaps/',
      cMapPacked: true,
    };
  }, []);

...

  return (
      <Document  options={opt}>
        {...}
      </Document>

  );
```

**Comment by Anhaachan** [comment]
> I only got this error when I resized the screen using the responsive dev tool, here is my call stack
> 
> ```
> Call Stack
> WorkerTransport.getOptionalContentConfig
> node_modules\pdfjs-dist\build\pdf.js (2481:0)
> PDFPageProxy.render
> node_modules\pdfjs-dist\build\pdf.js (1385:0)
> drawPageOnCanvas
> node_modules\react-pdf\dist\esm\Page\PageCanvas.js (68:0)
> commitHookEffectListMount
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (20002:0)
> commitHookPassiveMountEffects
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22055:0)
> commitPassiveMountOnFiber
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22160:0)
> recursivelyTraversePassiveMountEffects
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
> commitPassiveMountOnFiber
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22271:0)
> recursivelyTraversePassiveMountEffects
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
> commitPassiveMountOnFiber
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22271:0)
> recursivelyTraversePassiveMountEffects
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
> commitPassiveMountOnFiber
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22157:0)
> recursivelyTraversePassiveMountEffects
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
> commitPassiveMountOnFiber
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22271:0)
> recursivelyTraversePassiveMountEffects
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
> commitPassiveMountOnFiber
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22271:0)
> recursivelyTraversePassiveMountEffects
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
> commitPassiveMountOnFiber
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22157:0)
> recursivelyTraversePassiveMountEffects
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
> commitPassiveMountOnFiber
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22271:0)
> recursivelyTraversePassiveMountEffects
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
> commitPassiveMountOnFiber
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22271:0)
> recursivelyTraversePassiveMountEffects
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
> commitPassiveMountOnFiber
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22157:0)
> recursivelyTraversePassiveMountEffects
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
> commitPassiveMountOnFiber
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22271:0)
> recursivelyTraversePassiveMountEffects
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
> commitPassiveMountOnFiber
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22271:0)
> recursivelyTraversePassiveMountEffects
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
> commitPassiveMountOnFiber
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22157:0)
> recursivelyTraversePassiveMountEffects
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
> commitPassiveMountOnFiber
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22157:0)
> recursivelyTraversePassiveMountEffects
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
> commitPassiveMountOnFiber
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22271:0)
> recursivelyTraversePassiveMountEffects
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
> commitPassiveMountOnFiber
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22157:0)
> recursivelyTraversePassiveMountEffects
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
> commitPassiveMountOnFiber
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22271:0)
> recursivelyTraversePassiveMountEffects
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
> commitPassiveMountOnFiber
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22157:0)
> recursivelyTraversePassiveMountEffects
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
> commitPassiveMountOnFiber
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22157:0)
> recursivelyTraversePassiveMountEffects
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
> commitPassiveMountOnFiber
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22157:0)
> recursivelyTraversePassiveMountEffects
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
> commitPassiveMountOnFiber
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22157:0)
> recursivelyTraversePassiveMountEffects
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
> commitPassiveMountOnFiber
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22271:0)
> recursivelyTraversePassiveMountEffects
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22138:0)
> commitPassiveMountOnFiber
> node_modules\next\dist\compiled\react-dom\cjs\react-dom.development.js (22157:0)
> ```

Did you solved the error?

**Comment by vaibhavpras** [comment]
for anyone still facing this error, @illusdolphin and @jhunexjun 's answer worked for me for the latest version with a few edits.

from @jhunexjun's answer, i had to make 2 small changes

1.  extract the numPages from the document load event object as ({numPages}) instead

2.  add a state management check for document readiness before rendering the pages: basically <Page> was being rendered even before numPages was set upon document load resulting in the error. i am not sure if this is intended behavior for this component, but a quick fix using state management in the onDocumentLoadSuccess function helped me get past

```javascript
const [isDocumentReady, setIsDocumentReady] = useState(false);

function onDocumentLoadSuccess({numPages}) {
     setNumPages(numPages);
     setIsDocumentReady(true)
   }
.
.
.
return(
     <Document
           file={file}
           onLoadSuccess={onDocumentLoadSuccess}
           options={{
                cMapUrl: 'cmaps/',
                cMapPacked: true,
           }} >

             {isDocumentReady && Array.from(new Array(numPages), (el, index) => (
                  <Page key={`page_${index + 1}`} pageNumber={index + 1} />
              ))}

     </Document>
);
```

**Comment by c-oreills** [comment]
Thanks for all the help! Here's a similar solution that uses `useEffect` with the file prop as a dependency and a guard clause on `numPages` rather than adding extra state:

```
function RenderDocument({ file }): {
	const [numPages, setNumPages] = React.useState();
	
	React.useEffect(() => {
		setNumPages(undefined)
	}, [file]);

	function onDocumentLoadSuccess({ numPages: _numPages}) {
    	setNumPages(_numPages);
  	}

	return (
		<Document file={file} onLoadSuccess={onDocumentLoadSuccess}>
			{numPages != null &&
				[...new Array(numPages).keys()].map((pageIndex) => (
					<Page key={pageIndex} pageIndex={pageIndex} />
				))}
		</Document>
	);
}
```

**Comment by adem-labsi** [comment]
Hey everyone,

If anyone is experiencing this issue, I found a fix in **React** with **Vite** by removing **StrictMode** from **main.tsx**. It appears that the problem arises when the component renders twice. I'm using the example provided by the developers.

**Comment by johnnyshankman** [comment]
> for anyone still facing this error, @illusdolphin and @jhunexjun 's answer worked for me for the latest version with a few edits.
> 
> from @jhunexjun's answer, i had to make 2 small changes
> 
> 1. extract the numPages from the document load event object as ({numPages}) instead
> 2. add a state management check for document readiness before rendering the pages: basically `<Page>` was being rendered even before numPages was set upon document load resulting in the error. i am not sure if this is intended behavior for this component, but a quick fix using state management in the `onDocumentLoadSuccess` function helped me get past
> 
> ```js
> const [isDocumentReady, setIsDocumentReady] = useState(false);
> 
> function onDocumentLoadSuccess({numPages}) {
>      setNumPages(numPages);
>      setIsDocumentReady(true)
>    }
> .
> .
> .
> return(
>      <Document
>            file={file}
>            onLoadSuccess={onDocumentLoadSuccess}
>            options={{
>                 cMapUrl: 'cmaps/',
>                 cMapPacked: true,
>            }} >
> 
>              {isDocumentReady && Array.from(new Array(numPages), (el, index) => (
>                   <Page key={`page_${index + 1}`} pageNumber={index + 1} />
>               ))}
> 
>      </Document>
> );
> ```

This fixed my issue.

**Comment by will-thomson4** [comment]
I was still experiencing the issue with current version, but as some others suggested ^ fix worked for me by adding a unique key for the doc being loaded to the `<Document>` component

**Comment by LeviXIII** [comment]
> I was still experiencing the issue with current version, but as some others suggested ^ the fix that worked for me was adding a unique key for the doc being loaded to the `<Document>` component

This did it for me! Thanks will-thomson.

**Comment by chiragsonar** [comment]
> I was still experiencing the issue with current version, but as some others suggested ^ the fix that worked for me was adding a unique key for the doc being loaded to the `<Document>` component

@will-thomson4 Can you elaborate more on this? I tried the above solution, but when working with a large number of PDFs and navigating between them, I still encounter the error. Like what exactly worked for you?

**Comment by will-thomson4** [comment]
> > I was still experiencing the issue with current version, but as some others suggested ^ the fix that worked for me was adding a unique key for the doc being loaded to the `<Document>` component
> 
> [@will-thomson4](https://github.com/will-thomson4) Can you elaborate more on this? I tried the above solution, but when working with a large number of PDFs and navigating between them, I still encounter the error. Like what exactly worked for you?

@chiragsonar so, I load the PDFs in from a URL, and I just have:

```
  <Document
    key={fileUrl}
    file={fileUrl}
  >
```

And then for each new document it should re-render the` <Document/>` component

**Comment by Connoropolous** [comment]
same use case, same fix. react-pdf version (9.2.1) amazing thanks y'all!

**Comment by ssatriya** [comment]
If anyone still facing this issue, @will-thomson4 solution solve my problem with this sendWithPromise error. And I also remove this part of other solution

`Array.from(new Array(numPages), (el, index) =>`

which somehow make my last page of PDF duplicate.

- React-PDF 9.2.1
- Next.js 15.3.0

**Comment by huancnttuet** [comment]
The error occurs when the **file** prop changes rapidly, causing the PDF.js worker to be in an inconsistent state where it tries to access a null reference for sendWithPromise. This typically happens when:

1. Multiple PDF documents are being loaded/unloaded quickly
2. The component unmounts while a PDF is still loading
3. The PDF.js worker gets destroyed while operations are still pending

Here's the fix for this issue:

Add `key={fileUrl}`

```
<Document
      file={fileUrl}
      key={fileUrl} // Force remount when fileUrl changes
      ....
/>
```

**Comment by Chinaapanda** [comment]
@huancnttuet  it worked thank you so much

**Comment by xaho** [comment]
Sadly settings the key property does not seem to help when providing an ArrayBuffer instead of an URL.

**Comment by wojtekmaj** [comment]
@xaho When working with ArrayBuffer, you have two options:
- Storing `key` in state and incrementing it every time ArrayBuffer changes
- Creating a checksum out of ArrayBuffer data, for example like so:
  ```ts
  async function checksumSHA256(buffer: ArrayBuffer) {
    const hashBuffer = await crypto.subtle.digest("SHA-256", buffer);
  
    return Array.from(new Uint8Array(hashBuffer))
      .map((b) => b.toString(16).padStart(2, "0"))
      .join();
  }
  ```
  and using it as the key.

**Comment by joanfabregat** [comment]
I ran into the same issue when using a blob as a file. Using an identifier worked on the first load but failed on the second. I found a quick and dirty fix, inspired by @will-thomson4's solution:
```ts
<Document file={pdfBlob} key={crypto.randomUUID()}>
```

**Comment by 0rsa** [comment]
Same issue, and putting an unique key each time on Document, or waiting to be sure that the document is loaded before rendering the Page are not 100% working, the error can still appear

**Comment by magneto-09** [comment]
> Same issue, and putting an unique key each time on Document, or waiting to be sure that the document is loaded before rendering the Page are not 100% working, the error can still appear

Is there any fixed solution for it, especially when dealing with the blob object? Or are there any alternatives available?? @0rsa

**Comment by 0rsa** [comment]
I gave up and now use the pdf.js library directly without React components. It's less convenient but I no longer have this kind of problem.

**Comment by quynhduongphl** [comment]
I think by using [this comment](https://github.com/wojtekmaj/react-pdf/issues/974#issuecomment-3012437875), I've fixed my problem:
```
const [dataHash, setDataHash] = useState("hash");

useEffect(() => {
  const checksumSHA256 = async (buffer: ArrayBuffer) => {
    const hashBuffer = await crypto.subtle.digest("SHA-256", buffer);

    const hash = Array.from(new Uint8Array(hashBuffer))
      .map((b) => b.toString(16).padStart(2, "0"))
      .join();

    if (dataHash !== hash) {
      setDataHash(hash);
    }
  };

  if (data) {
    checksumSHA256(new Uint8Array(data).buffer);
  }
}, [data]);

const pdfData = useMemo(() => {
  if (dataHash) {
    return { data: data };
  }
  throw new Error("No data or file provided");
}, [dataHash, file]);

...

return (
  <Document
    file={pdfData}
    key={dataHash}
    onLoadSuccess={handleDocumentLoad}
    options={pdfOptions}
  >
    {Array.from(new Array(numPages), (_element, index) => (
      <Page
        key={`page_${index + 1}`}
        pageNumber={index + 1}
        width={containerWidth}
      />
    ))}
  </Document>
);
```

By anchoring the data array on its hash, this prevented react from continuously evaluating the data blob and causing worker issues.

**Comment by pramitchoudhary** [comment]
@quynhduongphl did you use the latest version or `9.2.1`?

**Comment by quynhduongphl** [comment]
> [@quynhduongphl](https://github.com/quynhduongphl) did you use the latest version of `9.2.1`?

latest version is `10.1.0` no?

**Comment by pramitchoudhary** [comment]
> > [@quynhduongphl](https://github.com/quynhduongphl) did you use the latest version of `9.2.1`?
> 
> latest version is `10.1.0` no?

Sorry, typo. Meant *latest or `9.2.1`?
Running similar issues on the Edge browser on Windows PC. In the process of upgrading.

**Comment by songkeys** [comment]
The issue also happens with [`<Activity>`](https://react.dev/reference/react/Activity).

To reproduce, wrap `<Document>` in `<Activity>`, then toggle the mode from `"visible"` to `"hidden"`, then back to `"visible"`.

```tsx
import { unstable_Activity as Activity, useState } from "react";
import { Document, Page } from "react-pdf";
import { pdfjs } from "react-pdf";

pdfjs.GlobalWorkerOptions.workerSrc = new URL(
  "pdfjs-dist/build/pdf.worker.min.mjs",
  import.meta.url
).toString();

export function PdfViewerDemo() {
  const [show, setShow] = useState(false);

  return (
    <div>
      <button onClick={() => setShow(!show)}>{show ? "Hide" : "Show"}</button>

      <Activity mode={show ? "visible" : "hidden"}>
        <Document file="...">
          <Page pageNumber={1} />
        </Document>
      </Activity>
    </div>
  );
}

```

**Comment by pramitchoudhary** [comment]
Forgot to update here, so passing files as blobs vs string worked for me.

**Comment by Meess** [comment]
Using 10.1.0, error also occured. As [huancnttuet](https://github.com/huancnttuet) said, it happens when the <Document/> "file" prop rapidly changes, while no key is given. In our case the file input is a Blob. Tried a few things:

- What worked (surprisingly), was adding the Blob in as key, not sure why as I expected react to convert keys to strings and therefore returning '[object Blob]' for each blob which would make it a static key (and that shouldn't work). As it works, react probably doesn't just call .toString() on it, but converts it in another way. We didn't choose this approach because if it does convert it in another way it means we have 10MB keys flying around, that can't be a good thing.
- Inspired by the Blob as key working (expecting it to be a static key, '[object Blob]',), adding just a static key 'someStaticString', but didn't work.
- Creating a hash of the blob and setting it as [quynhduongphl](https://github.com/quynhduongphl) suggested seems over complicated, but does work.

Eventually we went with incrementing the key when the blob changes, as suggested by https://github.com/wojtekmaj/react-pdf/issues/974#issuecomment-3012437875

i.e.:
``` typescript
const [pdfKey, setPdfKey] = useState(0);

useLayoutEffect(() => {
    setPdfKey((prevKey) => prevKey + 1);
}, [pdfBlob]);

<Document
    file={pdfBlob}
    key={pdfKey}
/>
```

**Comment by voidbass** [comment]
I fixed it with the code below, hope it helps everyone!

```
const [numPages, setNumPages] = useState<number>(0);
const onDocumentLoadSuccess = ({ numPages }: { numPages: number }): void => {
    setNumPages(numPages + 1);
  };

  <Document
              key={`${uploadedFileUrl}-${numPages}`}
              file={uploadedFileUrl}
              onLoadSuccess={onDocumentLoadSuccess}
            >
```

**Comment by Sergey-Shapovalov** [comment]
I'm not sure what exactly causes this problem, but I also experienced it in the situation when the file name changes before the document loading task completes. I solved it by assigning the key property of the <Document> to the file name like this:

`<Document key={fileName} file={fileName}/>`

Setting the key property forces React to unmount the Document object on key change and mount it again, cleaning up all the related stuff including the loading task.

**Comment by seedydocloop** [comment]
> Using 10.1.0, error also occured. As [huancnttuet](https://github.com/huancnttuet) said, it happens when the "file" prop rapidly changes, while no key is given. In our case the file input is a Blob. Tried a few things:
> 
> * What worked (surprisingly), was adding the Blob in as key, not sure why as I expected react to convert keys to strings and therefore returning '[object Blob]' for each blob which would make it a static key (and that shouldn't work). As it works, react probably doesn't just call .toString() on it, but converts it in another way. We didn't choose this approach because if it does convert it in another way it means we have 10MB keys flying around, that can't be a good thing.
> * Inspired by the Blob as key working (expecting it to be a static key, '[object Blob]',), adding just a static key 'someStaticString', but didn't work.
> * Creating a hash of the blob and setting it as [quynhduongphl](https://github.com/quynhduongphl) suggested seems over complicated, but does work.
> 
> Eventually we went with incrementing the key when the blob changes, as suggested by [#974 (comment)](https://github.com/wojtekmaj/react-pdf/issues/974#issuecomment-3012437875)
> 
> i.e.:
> 
> const [pdfKey, setPdfKey] = useState(0);
> 
> useLayoutEffect(() => {
>     setPdfKey((prevKey) => prevKey + 1);
> }, [pdfBlob]);
> 
> <Document
>     file={pdfBlob}
>     key={pdfKey}
> />

Can confirm this solved the bug when using blob as `file`.

In my use case, the error happens randomly when changing file.
It seems to be a timing issue with worker being destroyed while another one is being created.

I expected comparison of `file` instances to trigger too many effects, but it isn't triggered that often actually.

Running the following piece of code, I end up with `key` never bigger than `2`.

```
  const [key, setKey] = useState(0);

  console.log("key", key)

  useLayoutEffect(() => {
    if (file === undefined) {
      setKey(0);
    }
    setKey((prev) => prev + 1);
  }, [file]);

  return key;
```


Anyway, this hack is not very stable and feels like a short-term workaround.

@wojtekmaj 
> * Creating a checksum out of ArrayBuffer data, for example like so:
>   async function checksumSHA256(buffer: ArrayBuffer) {
>     const hashBuffer = await crypto.subtle.digest("SHA-256", buffer);
>   
>     return Array.from(new Uint8Array(hashBuffer))
>       .map((b) => b.toString(16).padStart(2, "0"))
>       .join();
>   }
>       
>         
>       
>   
>         
>       
>   
>       
>     
>   and using it as the key.

Your solution seems costly to simply generate a key, in the case of blobs or array buffers.
I'd love to see a real solution.

**Comment by teesien1998** [comment]
Hi @wojtekmaj, When will the recent commits (especially the promise cancellation fixes) be released as a new npm version? I see several commits from December but the latest npm version is still 10.2.0.

**Comment by korac** [comment]
@teesien1998 no work today, Merry Christmas man!

**Comment by wojtekmaj** [comment]
v10.3.0 is out.
